icu4c/source/test/intltest/localebuildertest.cpp

```diff
@@ -0,0 +1,1607 @@
+        "abcdefghi-12",  // First subtag too long
+        "a-ab-987654321",  // Third subtag too long
+        "987654321-a-0-3",  // First subtag too long
+    };
+    for (int i = 0; i < UPRV_LENGTHOF(ill_formed); i++) {
+        UErrorCode status = U_ZERO_ERROR;
+        LocaleBuilder bld;
+        bld.setExtension('x', ill_formed[i]);
+        Locale loc = bld.build(status);
+        if (status != U_ILLEGAL_ARGUMENT_ERROR) {
+            errln("setExtension('x', \"%s\") should fail but has no Error\n",
+                  ill_formed[i]);
+        }
+    }
+}
+
+void LocaleBuilderTest::TestSetExtensionOthers() {
+    LocaleBuilder bld;
+    bld.setLanguage("fr");
+    Verify(bld, "fr",
+           "setLanguage(\"fr\") should success but got Error: %s\n");
+
+    bld.setExtension('Z', "ab");
+    Verify(bld, "fr-z-ab",
+           "setExtension('Z', \"ab\") should success but got Error: %s\n");
+
+    bld.setExtension('0', "xyz12345-abcdefg");
+    Verify(bld, "fr-0-xyz12345-abcdefg-z-ab",
+           "setExtension('0', \"xyz12345-abcdefg\") should success but got Error: %s\n");
+
+    bld.setExtension('a', "01-12345678-ABcdef");
+    Verify(bld, "fr-0-xyz12345-abcdefg-a-01-12345678-abcdef-z-ab",
+           "setExtension('a', \"01-12345678-ABcdef\") should success but got Error: %s\n");
+
+    bld.setRegion("TH");
+    Verify(bld, "fr-TH-0-xyz12345-abcdefg-a-01-12345678-abcdef-z-ab",
+           "setRegion(\"TH\") should success but got Error: %s\n");
+
+    bld.setScript("Arab");
+    Verify(bld, "fr-Arab-TH-0-xyz12345-abcdefg-a-01-12345678-abcdef-z-ab",
+           "setRegion(\"Arab\") should success but got Error: %s\n");
+
+    bld.setExtension('A', "97");
+    Verify(bld, "fr-Arab-TH-0-xyz12345-abcdefg-a-97-z-ab",
+           "setExtension('a', \"97\") should success but got Error: %s\n");
+
+    bld.setExtension('a', "");
+    Verify(bld, "fr-Arab-TH-0-xyz12345-abcdefg-z-ab",
+           "setExtension('a', \"\") should success but got Error: %s\n");
+
+    bld.setExtension('0', "");
+    Verify(bld, "fr-Arab-TH-z-ab",
+           "setExtension('0', \"\") should success but got Error: %s\n");
+}
+
+void LocaleBuilderTest::TestSetExtensionValidateOthersWellFormed() {
+    static const char* well_formed[] = {
+        "ab",
+        "abc",
+        "abcefghi",
+        "01",
+        "234",
+        "87654321",
+        "0a-ab-87654321",
+        "87654321-ab-00-3A",
+    };
+    char ch = 'a';
+    for (int i = 0; i < UPRV_LENGTHOF(well_formed); i++, ch++) {
+        if (ch > 'z') ch = 'a';
+        UErrorCode status = U_ZERO_ERROR;
+        LocaleBuilder bld;
+        bld.setExtension(ch, well_formed[i]);
+        Locale loc = bld.build(status);
+        if (U_FAILURE(status)) {
+            errln("setExtension('%c', \"%s\") should success but got Error: %s\n",
+                  ch, well_formed[i], u_errorName(status));
+        }
+    };
+
+    // Test key with the whole the ASCII range.
+    for (char ch = 0x20; ch < 0x7F; ch++) {
```

Comment:
In EBCDIC, this would be 0x40..0xFF.
https://en.wikipedia.org/wiki/EBCDIC_037

I suggest you use a string literal with all letters and digits, and some other stuff, and iterate over that.